jxl/src/api/decoder.rs

```diff
@@ -1017,4 +1017,246 @@ pub(crate) mod tests {
+                            color_type,
+                            x,
+                            y,
+                            f32_val,
+                            u16_row[x],
+                            u16_val,
+                            error,
+                            use_simple
+                        );
+                    }
+                }
+            }
+        }
+    }
+
+    /// Test that f16 output matches f32 output within f16 precision tolerance.
+    #[test]
+    fn test_output_format_f16_matches_f32() {
+        use crate::api::{Endianness, JxlColorType, JxlDataFormat, JxlPixelFormat};
+        use crate::util::f16;
+
+        let file = std::fs::read("resources/test/conformance_test_images/bicycles.jxl").unwrap();
+
+        // Test both RGB and BGRA
+        for (color_type, num_samples) in [(JxlColorType::Rgb, 3), (JxlColorType::Bgra, 4)] {
+            let f32_format = JxlPixelFormat {
+                color_type,
+                color_data_format: Some(JxlDataFormat::f32()),
+                extra_channel_format: vec![],
+            };
+            let f16_format = JxlPixelFormat {
+                color_type,
+                color_data_format: Some(JxlDataFormat::F16 {
+                    endianness: Endianness::native(),
+                }),
+                extra_channel_format: vec![],
+            };
+
+            for use_simple in [true, false] {
+                let (f32_buffer, width, height) =
+                    decode_with_format::<f32>(&file, &f32_format, use_simple);
+                let (f16_buffer, _, _) = decode_with_format::<f16>(&file, &f16_format, use_simple);
+
+                // f16 has about 3 decimal digits of precision
+                // For values in [0,1], the relative error is about 0.001
+                let tolerance = 0.002;
+
+                for y in 0..height {
+                    let f32_row = f32_buffer.row(y);
+                    let f16_row = f16_buffer.row(y);
+                    for x in 0..(width * num_samples) {
+                        let f32_val = f32_row[x];
+                        let f16_val = f16_row[x].to_f32();
+                        let error = (f32_val - f16_val).abs();
+                        assert!(
+                            error < tolerance,
+                            "{:?} f16 mismatch at ({},{}): f32={}, f16={}, error={} (use_simple={})",
+                            color_type,
+                            x,
+                            y,
+                            f32_val,
+                            f16_val,
+                            error,
+                            use_simple
+                        );
+                    }
+                }
+            }
+        }
+    }
+
+    /// Helper function to decode an image with a specific format.
+    fn decode_with_format<T: crate::image::ImageDataType>(
```

Comment:
Shouldn't we use or modify `decode_image` instead?

For that matter, I feel like we might have duplicated and inlined `decode_image` a few times.